stl/inc/format

```diff
@@ -0,0 +1,2734 @@
+        return sizeof(bool);
+    case _Basic_format_arg_type::_Char_type:
+        return sizeof(_CharT);
+    case _Basic_format_arg_type::_Float_type:
+        return sizeof(float);
+    case _Basic_format_arg_type::_Double_type:
+        return sizeof(double);
+    case _Basic_format_arg_type::_Long_double_type:
+        return sizeof(long double);
+    case _Basic_format_arg_type::_Pointer_type:
+        return sizeof(void*);
+    case _Basic_format_arg_type::_CString_type:
+        return sizeof(const _CharT*);
+    case _Basic_format_arg_type::_String_type:
+        return sizeof(basic_string_view<_CharT>);
+    case _Basic_format_arg_type::_Custom_type:
+        return sizeof(void*) + sizeof(void (*)());
+    case _Basic_format_arg_type::_None:
+    default:
+        _STL_INTERNAL_CHECK(false);
+        return 0;
+    }
+}
+
+template <class _CharT, class _OutputIt>
+_NODISCARD _OutputIt _Fmt_write(_OutputIt _Out, monostate) {
+    _STL_INTERNAL_CHECK(false);
+    return _Out;
+}
+
+// This size is derived from the maximum length of an arithmetic type. The two contenders for widest are double and long
+// long. long long has a max length of ceil(log_10(2^64)) = 20 characters. double has a max length of
+// limits<double>::max_digits10 + the decimal + the sign + e + the exponent's sign + ceil(log_10(DBL_MAX_10_EXP))
+// = 17 + 1 + 1 + 1 + 3 = 24. An example is DBL_MAX which is "-1.7976931348623158e+308".
+inline constexpr size_t _Format_min_buffer_length = 24;
+
+// clang-format off
+template <class _CharT, class _OutputIt, class _Arithmetic>
+    requires(is_arithmetic_v<_Arithmetic> && !_CharT_or_bool<_Arithmetic, _CharT>)
+_NODISCARD _OutputIt _Fmt_write(_OutputIt _Out, const _Arithmetic _Value);
+// clang-format on
+
+template <class _CharT, class _OutputIt>
+_NODISCARD _OutputIt _Fmt_write(_OutputIt _Out, const bool _Value);
+
+template <class _CharT, class _OutputIt>
+_NODISCARD _OutputIt _Fmt_write(_OutputIt _Out, const _CharT _Value);
+
+template <class _CharT, class _OutputIt>
+_NODISCARD _OutputIt _Fmt_write(_OutputIt _Out, const void* const _Value);
+
+template <class _CharT, class _OutputIt>
+_NODISCARD _OutputIt _Fmt_write(_OutputIt _Out, const _CharT* _Value);
+
+template <class _CharT, class _OutputIt>
+_NODISCARD _OutputIt _Fmt_write(_OutputIt _Out, const basic_string_view<_CharT> _Value);
+
+#pragma warning(push)
+#pragma warning(disable : 4365) // 'argument': conversion from 'char' to 'const wchar_t', signed/unsigned mismatch
+// clang-format off
+template <class _CharT, class _OutputIt, class _Arithmetic>
+    requires (is_arithmetic_v<_Arithmetic> && !_CharT_or_bool<_Arithmetic, _CharT>)
+_NODISCARD _OutputIt _Fmt_write(_OutputIt _Out, const _Arithmetic _Value) {
+    // clang-format on
+    // TRANSITION, Reusable buffer
+    array<char, _Format_min_buffer_length> _Buffer;
+    const auto [_End, _Ec] = _STD to_chars(_Buffer.data(), _Buffer.data() + _Buffer.size(), _Value);

```

Comment:
I vaguely remember that we wanted to make this a plain array

Reply:
That could improve throughput and debug codegen slightly, but could reduce code clarity slightly. I'm not opposed as long as we centralize the built-in array's length (instead of repeating magic numbers), but I'm not sure we should do this.

Reply:
Well we already use `_Format_min_buffer_length`. The big advantage of the array is that it's impossible to have it decay without realizing it. Although I think it we use std::end()/std::size() it's not a huge deal.

Reply:
even if we don't centralize the length I'm OK changing to a plain array if we use std::end() and std::size() instead of directly using the size. Since those will fail if the array decays.